### PR TITLE
Enhancement: New BlockTracker and Json-Rpc-Engine based Provider

### PR DESCRIPTION
**Pull request by kumavis**
- [x] move to away from provider engine
  - [x] configure network via json-rpc-engine
  - [x] remove `provider._blockTracker` and `provider.on('block')` usage
  - [x] remove provider-engine dep
  - [x] remove docs mentioning provider-engine dep
- [x] move to `eth-block-tracker`@4
  - [x] update block event handlers
    - [x] eth-json-rpc-filters
    - [x] recent-block controller
    - [x] balance controller
    - [x] tx controller
    - [x] nonce-tracker
    - [x] others??
    - [x] fix missing block on block body requests
  - [x] update `eth-block-tracker` dep to use semver

bonus:
- [ ] add subs support
- [ ] dont subscribe to blocks unless necessary
- [ ] dedupe repeated code in controllers/network/createXyzClient
- [ ] publish `createBlockTrackerInspectorMiddleware`
- [x] nonce-tracker `_getCurrentBlock` -> `await getLatestBlock()`

### CODE REVIEW

**Comment by kumavis** [comment]
tests finally passing + consistently
unit tests became unstable for a bit, which was new to me
partially because of how nock is setup and reset, but metamask controller still exists and will poll against the reset nock

**Comment by kumavis** [comment]
despite tests passing, this is not ready and there are a few todos + unaddressed bugs in here

**Comment by kumavis** [comment]
image

seems like txs failing to confirm?

**Comment by kumavis** [comment]
image
also seems like nonce is bad

**Comment by kumavis** [comment]
keep seeing this error in e2e tests:
> "[ethjs-rpc] rpc error with payload {\"id\":3743105454522,\"jsonrpc\":\"2.0\",\"params\":[\"0xe18035bf8712672935fdb4e5e431b1a0183d2dfc\",\"0x56e9fb\"],\"method\":\"eth_getTransactionCount\"} Error: LevelUpArrayAdapter named 'blocks' index out of range: index 5695995; length: 1"

this ganache internal error message means the block doesnt exist, reported that they should handle that more gracefully
ths problem is we're hitting ganache with a mainnet block number
may want to modify e2e tests to inject some sort of config before they run

**Comment by kumavis** [comment]
```
1) Restore from Seed (logout and logging back in works at this point)
2) Create new account
3) Logout
4) “Incorrect password” at login
```

**Comment by kumavis** [comment]
merge conflict resolution removed the error-stack simplification (e.g. not giving internal error stacks to dapps on rejected tx)

**Comment by bdresser** [comment]
let's revisit #1999 once this is merged

**Comment by kumavis** [comment]
after some fixes, passed my informal QA

File: app/scripts/metamask-controller.js [hunk 1]
```diff
@@ -233,28 +233,22 @@ module.exports = class MetamaskController extends EventEmitter {
...
+        } else {
+          return []
         }
-        cb(null, result)
       },
       // tx signing
-      // old style msg signing
-      processMessage: this.newUnsignedMessage.bind(this),
-      // personal_sign msg signing
+      processTransaction: this.newUnapprovedTransaction.bind(this),
```
Comment (frankiebee) [inline app/scripts/metamask-controller.js]: was this fixed so that it now passes the request object?

Comment (frankiebee) [reply app/scripts/metamask-controller.js]: checked; it's fine

**Comment by bdresser** [comment]
Couple weeks back we grouped a bunch of issues that were _possibly_ going to be addressed by this PR. @frankiebee @kumavis would you mark off what's handled and I'll leave the rest open? 

- [ ] Network: disable cache on localhost #3088
- [ ] Reduce load on infura #3143
- [ ] Filters/Events/Logs not working correctly #2978
- [ ] will not start! stuck on "connecting to network" #2636
- [ ] Detect when TestRPC changes & restart provider #2373
- [ ] Add websocket support #1645
- [ ] Pass `origin` HTTP header on requests to Infura #4739

**Review by danfinlay — APPROVED** [review]
Let's begin the process of trying this out on develop!

**Comment by kumavis** [comment]
@bdresser thanks for that list! some of them may be complete but are not QA'd

**Comment by saturnial** [comment]
Hi there! When can we expect this fix to be available to the public?

We're currently experiencing issues where our calls to read data from the blockchain are hitting the cache -- preventing us from seeing updated data when transactions are newly mined.

We believe this to be caused by the caching policy embedded in MetaMask but would love to confirm this...

Thanks!

**Comment by danfinlay** [comment]
We're hoping to start rolling this branch out this week, but it depends on the number of bugs our team finds (we have it on `develop` branch now).

**Comment by tarekskr** [comment]
Hey @danfinlay , has this been rolled out yet? Any ETA? Thanks a lot!

**Comment by bdresser** [comment]
@tarekskr still testing on develop, will likely roll out early next week!

**Comment by tarekskr** [comment]
@bdresser this is really awesome, can't wait!!

**Pull request merged**